File: jp-livedata-java/src/main/java/com/mao/jp_livedata_java/core/BaseBusObserverWrapper.java
package com.mao.jp_livedata_java.core;

import androidx.lifecycle.BusLiveData;
import androidx.lifecycle.Observer;

import com.mao.jp_livedata_java.Logger;

/**
 * @author zhangkun
 * @time 2022/2/16 17:23
 * @Description
 */
abstract public class BaseBusObserverWrapper<T> implements Observer<T> {

    private final BusLiveData<T> mBusLiveData;

    private final Observer<T> mObserver;

    private final int mLastVersion;

    private final String TAG = "BaseBusObserverWrapper";

    public BaseBusObserverWrapper(Observer<T> mObserver, BusLiveData<T> mBusLiveData) {
        this.mObserver = mObserver;
        this.mBusLiveData = mBusLiveData;
        this.mLastVersion = mBusLiveData.getVersion();
    }


    @Override
    public void onChanged(T t) {
        if (mLastVersion >= mBusLiveData.getVersion()) {
            // LiveData 的版本号没有更新过，说明并没有新数据，只是因为
            // 当前Observer的版本号比 LiveData 低导致的调用 onChange()
            return;
        }
        try {
            mObserver.onChanged(t);
        } catch (Exception e) {
            e.printStackTrace();
            Logger.e(TAG, "error on Observer onChanged() = " + e.getMessage());
        }
    }
}

File: jp-livedata/src/main/java/com/mao/jp_livedata/core/BaseBusObserverWrapper.kt
package com.mao.jp_livedata.core


import androidx.lifecycle.BusLiveData
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.Observer
import com.mao.jp_livedata.Logger

internal open class BaseBusObserverWrapper<T>(private val mObserver: Observer<in T>, private val mBusLiveData: BusLiveData<T>) : Observer<T> {

    private val mLastVersion = mBusLiveData.version

    private val TAG = "BaseBusObserverWrapper"

    override fun onChanged(t: T?) {
        Logger.d(TAG,"msg receiver = " + t.toString())
        if (mLastVersion >= mBusLiveData.version){
            // LiveData 的版本号没有更新过，说明并没有新数据，只是因为
            // 当前Observer的版本号比 LiveData 低导致的调用 onChange()
            return
        }
        try {
            mObserver.onChanged(t)
        }catch (e:Exception){
            Logger.e(TAG,"error on Observer onChanged() = " + e.message)
        }
    }
}

